pkg/storage/stores/local/shipper.go

```diff
@@ -286,11 +269,18 @@ func (s *Shipper) forEach(ctx context.Context, period string, callback func(db *
 		} else {
 			level.Info(pkg_util.Logger).Log("msg", fmt.Sprintf("downloading all files for period %s", period))
 
-			fc = &filesCollection{files: map[string]downloadedFiles{}}
+			fc = newFilesCollection()
 			s.downloadedPeriods[period] = fc
 			s.downloadedPeriodsMtx.Unlock()
 
-			if err := s.downloadFilesForPeriod(ctx, period, fc); err != nil {
+			// Using background context to avoid cancellation of download when request times out.
+			// We would anyways need the files for serving next requests.
+			if err := s.downloadFilesForPeriod(context.Background(), period, fc); err != nil {
```

Comment:
You should use another go routine for downloading files and wait for it unless the context is cancelled.

Could be a background goroutine that is waiting for those requests. The problem with using the current goroutine is that even though you will return in case of timeout, the goroutine here will stay stuck. In case of an issue you could starve the http goroutines with this.